src/cmake/Findtinyxml2.cmake

```diff
@@ -0,0 +1,37 @@
+# Copied from https://github.com/ros2/tinyxml2_vendor/blob/master/cmake/Modules/FindTinyXML2.cmake
+
+# TinyXML2_FOUND
+# TinyXML2_INCLUDE_DIRS
+# TinyXML2_LIBRARIES
+
+# try to find the CMake config file for TinyXML2 first
+find_package(TinyXML2 CONFIG QUIET)
+if(TinyXML2_FOUND)
+  message(STATUS "Found TinyXML2 via Config file: ${TinyXML2_DIR}")
+  if(NOT TINYXML2_LIBRARY)
+    # in this case, we're probably using TinyXML2 version 5.0.0 or greater
+    # in which case tinyxml2 is an exported target and we should use that
+    if(TARGET tinyxml2)
+      set(TINYXML2_LIBRARY tinyxml2)
+    elseif(TARGET tinyxml2::tinyxml2)
+      set(TINYXML2_LIBRARY tinyxml2::tinyxml2)
+    endif()
+  endif()
+else()
+  find_path(TINYXML2_INCLUDE_DIR NAMES tinyxml2.h)
```

Comment:
Isn't this exactly the thing of scanning paths first before trying pkg_config? Please don't, use pkgconfig in preference.

Reply:
Yes, but it depends on `CMAKE_FIND_ROOT_PATH`. And that should be set by our toolchain. I trust you if you say that it doesn't work, but I'd be happy to experiment a bit with the cross-compilation setup :+1:.

And/or we can add pkgconfig before, of course :+1:.

> The CMake variable CMAKE_FIND_ROOT_PATH specifies one or more directories to be prepended to all other search directories. This effectively “re-roots” the entire search under given locations. Paths which are descendants of the CMAKE_STAGING_PREFIX are excluded from this re-rooting, because that variable is always a path on the host system. By default the CMAKE_FIND_ROOT_PATH is empty.

Reply:
Well I cannot tell you an exact failure scenario, but... 
1. I doubt this scales having to write such "miniprograms" for every library we use, and it is absolutely no fun dealing with them one by one if they are snowflakes. Please figure *one* canonical way of doing it and use it throughout.
2. If always approached in this ad-hoc fashion, we will have different search scripts even for the same library in different build targets. (Actually, we have that already). Not fun.
3. (more philosophical) We used to work a decade with gnu autotools trying to locate libraries by probing random things in "supposedly well-known paths" (which sucked) but switched over to not "guessing" anymore somewhere 15-20 years ago by having "authoritative" package description files. Either enough of the reasons why "guessing sucks" are not valid anymore, or we just see the cycle of the next generation of tool implementors having to relearn the same things. The thing that is a particular mystery to me is why if "find_path" works so well you would even bother with "find_package" for the cmake package definition file -- since pkg-config and cmake packages are semantically equivalent I assume the reason is more "dislike" than technical soundness.


use one kind of authoritative package definition if it exists (cmake)the (authoritative) cmake-package definition from system if it is there (good)

Reply:
1. There seems to be no canonical way given that we also need to build on Windows. Hence `pkgconfig` is not a solution.
2. I want to learn if and how these scripts work. I don't want to avoid them out of fear.
3. I would guess that using `find_package` you can use something specific if required rather than the system one.

Reply:
What's funny though is that there seems to be a `Findjsoncpp.cmake` file on my Fedora but then there is none available on my Arch Linux at home. :thinking: @JonasVautherin .

Reply:
> 1. There seems to be no canonical way given that we also need to build on Windows. Hence `pkgconfig` is not a solution.

There is a canonical way: 1. Try cmake config. 2. Try pkg-config. 3. Search. 

In your specific case, the first step is going to fail everywhere except as in superbuild. The second step is going to succeed everywhere except windows. I am not sure that the last step stands a chance of succeeding anywhere if the first two did not (only possibility would be a system where the library was "hand-installed" without either of the two config files), but if it does then most likely it has definitely produced a very wrong result.

What I do not like is the approach of "open-coding" this for every library.

> 2. I want to learn if and how these scripts work. I don't want to avoid them out of fear.
> 3. I would guess that using `find_package` you can use something specific if required rather than the system one.

Yes that works correctly. (As does pkgconfig). find_path is the thing doing the guessing.